From 217f86cf1771800a0e390c50470a784e32b35e7c Mon Sep 17 00:00:00 2001
From: Jing Liu <jing2.liu@intel.com>
Date: Thu, 21 Oct 2021 03:23:57 -0700
Subject: [PATCH 0696/1418] x86/fpu: Change xstate_get_guest_group_perm() to
 external

To avoid weird guest failures at runtime due to permission failures of
a reallocation request, KVM checks the permissions of the vcpu for the
extend features at kvm_vcpu_ioctl_set_cpuid2() time, when QEMU wants to
advertise the extended features (e.g. AMX) for the first time.

Thus, KVM also need the API to get guest permission. Change
xstate_get_guest_group_perm() to external for KVM.

Signed-off-by: Jing Liu <jing2.liu@intel.com>
---
 arch/x86/include/asm/fpu/api.h | 2 ++
 arch/x86/kernel/fpu/xstate.c   | 5 +++++
 arch/x86/kernel/fpu/xstate.h   | 5 -----
 3 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/arch/x86/include/asm/fpu/api.h b/arch/x86/include/asm/fpu/api.h
index 6053674f9132..7532f73c82a6 100644
--- a/arch/x86/include/asm/fpu/api.h
+++ b/arch/x86/include/asm/fpu/api.h
@@ -138,6 +138,8 @@ static inline void fpstate_free(struct fpu *fpu) { }
 /* fpstate-related functions which are exported to KVM */
 extern void fpstate_clear_xstate_component(struct fpstate *fps, unsigned int xfeature);
 
+extern inline u64 xstate_get_guest_group_perm(void);
+
 /* KVM specific functions */
 extern bool fpu_alloc_guest_fpstate(struct fpu_guest *gfpu);
 extern void fpu_free_guest_fpstate(struct fpu_guest *gfpu);
diff --git a/arch/x86/kernel/fpu/xstate.c b/arch/x86/kernel/fpu/xstate.c
index 3e9709969e61..6bbfb311f9c8 100644
--- a/arch/x86/kernel/fpu/xstate.c
+++ b/arch/x86/kernel/fpu/xstate.c
@@ -1757,6 +1757,11 @@ static inline int xstate_request_perm(unsigned long idx, bool guest)
 }
 #endif  /* !CONFIG_X86_64 */
 
+inline u64 xstate_get_guest_group_perm(void)
+{
+	return xstate_get_group_perm(true);
+}
+EXPORT_SYMBOL_GPL(xstate_get_guest_group_perm);
 /**
  * fpu_xstate_prctl - xstate permission operations
  * @tsk:	Redundant pointer to current
diff --git a/arch/x86/kernel/fpu/xstate.h b/arch/x86/kernel/fpu/xstate.h
index b62e6ff38520..7ccc63a68847 100644
--- a/arch/x86/kernel/fpu/xstate.h
+++ b/arch/x86/kernel/fpu/xstate.h
@@ -34,11 +34,6 @@ static inline u64 xstate_get_host_group_perm(void)
 	return xstate_get_group_perm(false);
 }
 
-static inline u64 xstate_get_guest_group_perm(void)
-{
-	return xstate_get_group_perm(true);
-}
-
 enum xstate_copy_mode {
 	XSTATE_COPY_FP,
 	XSTATE_COPY_FX,
-- 
2.31.1

